src/core/mac/sub_mac.cpp

```diff
@@ -500,6 +500,19 @@ void SubMac::HandleTransmitDone(TxFrame &aFrame, RxFrame *aAckFrame, Error aErro
     bool ccaSuccess = true;
     bool shouldRetx;
 
+#if OPENTHREAD_RADIO
+    if (aFrame.GetChannel() != aFrame.GetRxChannelAfterTxDone() && mRxOnWhenIdle)
+    {
+        // On RCP build, we switch immediately to the specified RX
+        // channel if it is different from the channel on which frame
+        // was sent. On FTD or MTD builds we don't need to do
+        // the same as the `Mac` will switch the channel from the
+        // `mCallbacks.TransmitDone()`.
+
+        IgnoreError(Get<Radio>().Receive(aFrame.GetRxChannelAfterTxDone()));
+    }
+#endif
```

Comment:
This PR seems to move the check from below after `SetState(kStateReceive)` to here the top of the function. I suggest keeping it there?

If `SubMac` decides to retransmit the frame, we don't want to change the channel. In particular, if `SubMac` is doing CSMA, it should stay on the frame tx channel).

Reply:
I thought it may be useful to update the `radio.h` documentation to mention retries (help platform developers), but it looks like we already mention it: 🙂

[https://github.com/openthread/openthread/blob/358b0a3e18b9d5e3b5ae48f55de0069141086f4c/include/openthread/platform/radio.h#L294-L299](https://github.com/openthread/openthread/blob/358b0a3e18b9d5e3b5ae48f55de0069141086f4c/include/openthread/platform/radio.h#L294-L299)

Reply:
The `RxChannelAfterTxDone` is mostly the Thread network's main operational channel. And the minimal expected backoff time for retries is `3.5 ms`. IIUC, switching back to the main channel increases the actual duty cycle in the main channel, so it can potentially reduce packet loss in the main channel. I reverted the moving and focus on rx-on-when-idle in this PR only. I'll submit a separate if you also agree with applying the RxChannelAfterTxDone between retries.

Reply:
Sounds good.

Generally I think requiring to switch to rx channel (if different than tx frame) in between tx retries would have side-effects. Will need more thoughts and can possibly complicate things.